File: shared/src/main/java/com/runtracker/shared/data/model/PersonalBest.kt
package com.runtracker.shared.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * Represents a personal best record for a specific distance
 */
@Entity(tableName = "personal_bests")
data class PersonalBest(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    
    // Standard distances in meters
    val distanceMeters: Int,
    
    // Fastest time records
    val fastestTimeMillis: Long? = null,
    val fastestTimeRunId: Long? = null,
    val fastestTimeDate: Long? = null,
    val fastestTimePaceSecondsPerKm: Double? = null,
    
    // Lowest average heart rate records (for same distance)
    val lowestAvgHeartRate: Int? = null,
    val lowestHrRunId: Long? = null,
    val lowestHrDate: Long? = null,
    val lowestHrTimeMillis: Long? = null,
    
    // Workout type for this PB (optional - for workout-specific PBs)
    val workoutType: WorkoutType? = null
) {
    val distanceKm: Double get() = distanceMeters / 1000.0
    
    val fastestTimeFormatted: String get() {
        val millis = fastestTimeMillis ?: return "--:--"
        val totalSeconds = millis / 1000
        val hours = totalSeconds / 3600
        val minutes = (totalSeconds % 3600) / 60
        val seconds = totalSeconds % 60
        return if (hours > 0) {
            String.format("%d:%02d:%02d", hours, minutes, seconds)
        } else {
            String.format("%d:%02d", minutes, seconds)
        }
    }
    
    val fastestPaceFormatted: String get() {
        val pace = fastestTimePaceSecondsPerKm ?: return "--:--"
        val minutes = (pace / 60).toInt()
        val seconds = (pace % 60).toInt()
        return String.format("%d:%02d", minutes, seconds)
    }
    
    companion object {
        // Standard distances in meters
        const val DISTANCE_1K = 1000
        const val DISTANCE_5K = 5000
        const val DISTANCE_10K = 10000
        const val DISTANCE_HALF_MARATHON = 21097
        const val DISTANCE_MARATHON = 42195
        
        val STANDARD_DISTANCES = listOf(
            DISTANCE_1K,
            DISTANCE_5K,
            DISTANCE_10K,
            DISTANCE_HALF_MARATHON,
            DISTANCE_MARATHON
        )
        
        fun getDistanceName(distanceMeters: Int): String {
            return when (distanceMeters) {
                DISTANCE_1K -> "1K"
                DISTANCE_5K -> "5K"
                DISTANCE_10K -> "10K"
                DISTANCE_HALF_MARATHON -> "Half Marathon"
                DISTANCE_MARATHON -> "Marathon"
                else -> "${distanceMeters / 1000.0}km"
            }
        }
    }
}

/**
 * Represents the current run progress compared to a personal best
 */
data class CompeteProgress(
    val distanceMeters: Int,
    val personalBestTimeMillis: Long,
    val currentTimeMillis: Long,
    val currentDistanceMeters: Double,
    val pbDistanceAtCurrentTime: Double, // Where PB runner would be at current time
    val timeDifferenceMillis: Long, // Positive = behind PB, Negative = ahead of PB
    val isAheadOfPb: Boolean
) {
    val timeDifferenceFormatted: String get() {
        val absMillis = kotlin.math.abs(timeDifferenceMillis)
        val totalSeconds = absMillis / 1000
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        val sign = if (isAheadOfPb) "-" else "+"
        return if (minutes > 0) {
            "$sign${minutes}:${String.format("%02d", seconds)}"
        } else {
            "$sign${seconds}s"
        }
    }
    
    // Progress as percentage (0.0 to 1.0)
    val progressPercent: Double get() = (currentDistanceMeters / distanceMeters).coerceIn(0.0, 1.0)
    val pbProgressPercent: Double get() = (pbDistanceAtCurrentTime / distanceMeters).coerceIn(0.0, 1.0)
}

File: iosApp/GoSteady/Sources/Models/PersonalBest.swift
import Foundation
import SwiftData

@Model
final class PersonalBest {
    @Attribute(.unique) var id: UUID
    var distanceMeters: Int
    var fastestTimeMillis: Int64?
    var fastestTimeRunId: UUID?
    var fastestTimeDate: Date?
    var fastestTimePaceSecondsPerKm: Double?
    var lowestAvgHeartRate: Int?
    var lowestHrRunId: UUID?
    var lowestHrDate: Date?
    var lowestHrTimeMillis: Int64?
    var workoutType: WorkoutType?

    // Gym exercise personal bests
    var exerciseId: UUID?
    var exerciseName: String?
    var bestWeight: Double
    var bestReps: Int
    var estimatedOneRepMax: Double
    var achievedDate: Date

    var distanceKm: Double { Double(distanceMeters) / 1000.0 }

    var fastestTimeFormatted: String {
        guard let millis = fastestTimeMillis else { return "--:--" }
        let totalSeconds = millis / 1000
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }

    var fastestPaceFormatted: String {
        guard let pace = fastestTimePaceSecondsPerKm else { return "--:--" }
        let minutes = Int(pace / 60)
        let seconds = Int(pace.truncatingRemainder(dividingBy: 60))
        return String(format: "%d:%02d", minutes, seconds)
    }

    static let distance1K = 1000
    static let distance5K = 5000
    static let distance10K = 10000
    static let distanceHalfMarathon = 21097
    static let distanceMarathon = 42195

    static let standardDistances = [distance1K, distance5K, distance10K, distanceHalfMarathon, distanceMarathon]

    static func getDistanceName(_ distanceMeters: Int) -> String {
        switch distanceMeters {
        case distance1K: return "1K"
        case distance5K: return "5K"
        case distance10K: return "10K"
        case distanceHalfMarathon: return "Half Marathon"
        case distanceMarathon: return "Marathon"
        default: return "\(Double(distanceMeters) / 1000.0)km"
        }
    }

    init(
        id: UUID = UUID(),
        distanceMeters: Int = 0,
        fastestTimeMillis: Int64? = nil,
        fastestTimeRunId: UUID? = nil,
        fastestTimeDate: Date? = nil,
        fastestTimePaceSecondsPerKm: Double? = nil,
        lowestAvgHeartRate: Int? = nil,
        lowestHrRunId: UUID? = nil,
        lowestHrDate: Date? = nil,
        lowestHrTimeMillis: Int64? = nil,
        workoutType: WorkoutType? = nil,
        exerciseId: UUID? = nil,
        exerciseName: String? = nil,
        bestWeight: Double = 0,
        bestReps: Int = 0,
        estimatedOneRepMax: Double = 0,
        achievedDate: Date = Date()
    ) {
        self.id = id
        self.distanceMeters = distanceMeters
        self.fastestTimeMillis = fastestTimeMillis
        self.fastestTimeRunId = fastestTimeRunId
        self.fastestTimeDate = fastestTimeDate
        self.fastestTimePaceSecondsPerKm = fastestTimePaceSecondsPerKm
        self.lowestAvgHeartRate = lowestAvgHeartRate
        self.lowestHrRunId = lowestHrRunId
        self.lowestHrDate = lowestHrDate
        self.lowestHrTimeMillis = lowestHrTimeMillis
        self.workoutType = workoutType
        self.exerciseId = exerciseId
        self.exerciseName = exerciseName
        self.bestWeight = bestWeight
        self.bestReps = bestReps
        self.estimatedOneRepMax = estimatedOneRepMax
        self.achievedDate = achievedDate
    }
}

struct CompeteProgress: Codable, Hashable {
    var distanceMeters: Int
    var personalBestTimeMillis: Int64
    var currentTimeMillis: Int64
    var currentDistanceMeters: Double
    var pbDistanceAtCurrentTime: Double
    var timeDifferenceMillis: Int64
    var isAheadOfPb: Bool

    var timeDifferenceFormatted: String {
        let absMillis = abs(timeDifferenceMillis)
        let totalSeconds = absMillis / 1000
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        let sign = isAheadOfPb ? "-" : "+"
        if minutes > 0 {
            return "\(sign)\(minutes):\(String(format: "%02d", seconds))"
        } else {
            return "\(sign)\(seconds)s"
        }
    }

    var progressPercent: Double {
        min(max(currentDistanceMeters / Double(distanceMeters), 0.0), 1.0)
    }

    var pbProgressPercent: Double {
        min(max(pbDistanceAtCurrentTime / Double(distanceMeters), 0.0), 1.0)
    }
}
